### PR TITLE
ci: implement lint CI workflow and document the style

### LINKED ISSUE #124: Document style guide
**Device type**
- [x] ESP32
- [x] ESP8226

**Framework version**
- [x] ESP-IDF latest master
- [x] ESP-IDF 3.2.x
- [x] ESP-IDF 3.3.x
- [x] ESP8266 RTOS SDK latest master
- [x] ESP8266 RTOS SDK 3.2

**Describe the bug**

code style is not documented. the repository has `.astylerc` and the documentation should mention it.

### PR DESCRIPTION
**Pull request by trombik**
fixes #124, #184

instead of `astyle`, which has not been actively maintained, `clang-format` and its wrapper are used to lint the code.

@UncleRus I think `astylerc` in the repository has been ported properly (it does not change anything `components/example/example.c`). as `clang-format` has more options than `astyle`, please have a look at the provided example code. if you find anything you don't like, let me know an example to demonstrate your style.

### CODE REVIEW

File: .clang-format [hunk 1]
```diff
@@ -0,0 +1,23 @@
+---
+# to see all the options in the `BasedOnStyle`, use `clang-format --dump-config`.
+# when you add a new option, make sure `comonents/example` has a valid style
+# that conforms the new option.
+
+BasedOnStyle: LLVM
+
+BreakBeforeBraces: Custom
+BraceWrapping:
+    AfterClass: false
```
Comment (UncleRus) [inline .clang-format:10]: Also true. We don't use cpp yet, but who knows

File: .clang-format [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    AfterNamespace: true
+    AfterStruct: true
+    AfterUnion: true
+    AfterExternBlock: false
+    BeforeCatch: true
+    BeforeElse: true
+    IndentBraces: false
+IndentWidth: 4
+IndentCaseLabels: true
+UseCRLF: true
```
Comment (UncleRus) [inline .clang-format:23]: False. Unix-style line endings only.

Comment (UncleRus) [reply .clang-format:23]: Also `SpaceBeforeParens` must be set to `ControlStatements`

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -200,6 +201,47 @@ Note that `git rebase` rewrites the commit history. You should avoid `git
...
 additional steps to ensure the review result is included.
 
+### Code style
+
+We use a style for source files based on [LLVM Coding Standards](https://llvm.org/docs/CodingStandards.html)
+except some cases, notably brace wrapping. Here is a brief list of the styles.
+
+- Use `snake_case`, not `CamelCase`
+- Use spaces. The indent width is four
+- Use `\r\n`, or `CR + LF` for line breaks
```
Comment (UncleRus) [inline CONTRIBUTING.md:211]: No-no, Unix-style, `\r`

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -200,6 +201,47 @@ Note that `git rebase` rewrites the commit history. You should avoid `git
...
+- Use `\\` for inline comments. Use `/* */` for multi line comments after an
```
Comment (UncleRus) [inline CONTRIBUTING.md:212]: wrong slashes, must be `//`

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -200,6 +201,47 @@ Note that `git rebase` rewrites the commit history. You should avoid `git
...
+  empty line
+- Break before braces in *most cases* (functions, conditionals, control
+  statements, etc)
+- Always check given arguments
+- Always check return code, return value, or `errno`
+- Return `esp_err_t` from functions where possible
+- Document public functions, data types, and macros in header files
+
```
Comment (UncleRus) [inline CONTRIBUTING.md:220]: - For type names use the suffix `_t`: `typedef int my_type_t`, for function types `_cb_t`: `void (*my_callback_cb_t)()`, for structs `_s`: `struct my_struct_s`. Type names must be in snake-case too.
- Macros must be in `UPPER_CASE`

File: .clang-format (hunk 2 above)
Comment (UncleRus) [reply .clang-format:23]: And `PointerAlignment` must be `Right`

Comment (trombik) [reply .clang-format:23]: fixed.

as SpaceBeforeParens = SpaceBeforeParens, and PointerAlignment = Right in LLVM style, they are omitted.

File: CONTRIBUTING.md (hunk 3 above)
Comment (trombik) [reply CONTRIBUTING.md]: fixed

File: CONTRIBUTING.md (hunk 4 above)
Comment (trombik) [reply CONTRIBUTING.md]: fixed

File: CONTRIBUTING.md (hunk 5 above)
Comment (trombik) [reply CONTRIBUTING.md:224]: fixed

**Comment by UncleRus** [comment]
CI build failed because of my error.
Merged.

Thank you!

**Pull request merged**